File: src/main/java/ninja/hudy/infosmog/api/v1/to/Converter.java
package ninja.hudy.infosmog.api.v1.to;

import ninja.hudy.infosmog.model.CountyRegion;
import ninja.hudy.infosmog.model.Measurement;
import ninja.hudy.infosmog.model.MeasurementType;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

public final class Converter {

    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    private static final String ZONE_ID = "Europe/Warsaw";

    private Converter() {
        // Utils class
    }

    public static MeasurementTO getMeasurementTO(Measurement measurement, Integer hour) {
        MeasurementTO measurementTO = new MeasurementTO();

        measurementTO.setTimestamp(measurement.getTimestamp());
        measurementTO.setValue(measurement.getValue());
        measurementTO.setHour(hour);

        return measurementTO;
    }

    public static MeasurementTypeTO getMeasurementTypeTO(MeasurementType measurementType, List<MeasurementTO> measurementTOs) {
        MeasurementTypeTO measurementTypeTO = new MeasurementTypeTO();

        measurementTypeTO.setId(measurementType.getId());
        measurementTypeTO.setName(measurementType.getName());
        measurementTypeTO.setDisplayName(measurementType.getDisplayName());
        measurementTypeTO.setMeasurements(measurementTOs);
        measurementTypeTO.setThresholds(getThresholdTO(measurementType));

        return measurementTypeTO;
    }

    public static ThresholdTO getThresholdTO(MeasurementType measurementType) {
        ThresholdTO thresholdTO = new ThresholdTO();

        thresholdTO.setAcceptable(measurementType.getAcceptable());
        thresholdTO.setInformative(measurementType.getInformative());
        thresholdTO.setAlarm(measurementType.getAlarm());

        return thresholdTO;
    }

    public static CountyTO getCountyTO(CountyRegion county, List<MeasurementTypeTO> measurementTypeTOs, ZonedDateTime dateTime) {
        CountyTO countyTO = new CountyTO();
        String now = dateTime.withZoneSameInstant(ZoneId.of(ZONE_ID)).format(DATE_TIME_FORMATTER);

        countyTO.setName(county.getName());
        countyTO.setNow(now);
        countyTO.setWeatherType(null == county.getWeatherType() ? 0 : county.getWeatherType());
        countyTO.setWeatherDegree(null == county.getWeatherDegree() ? 0 : Math.round(county.getWeatherDegree()));

        countyTO.setMeasurementTypes(measurementTypeTOs);

        return countyTO;
    }

    public static CountyLightTO getCountyLightTO(CountyRegion county) {
        CountyLightTO countyLightTO = new CountyLightTO();

        countyLightTO.setName(county.getName());
        countyLightTO.setId(county.getId());
        countyLightTO.setVoivodeship(county.getSuperName());

        return countyLightTO;
    }
}

File: src/main/kotlin/pl/org/smok/infosmog/api/v1/to/Converter.kt
package pl.org.smok.infosmog.api.v1.to

import pl.org.smok.infosmog.model.CountyRegion
import pl.org.smok.infosmog.model.Measurement
import pl.org.smok.infosmog.model.MeasurementType

import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

object Converter {
    private val DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
    private val ZONE_ID = "Europe/Warsaw"

    fun getMeasurementTO(measurement: Measurement, hour: Int?): MeasurementTO {
        val measurementTO = MeasurementTO()

        measurementTO.timestamp = measurement.timestamp
        measurementTO.value = measurement.value
        measurementTO.hour = hour

        return measurementTO
    }

    fun getMeasurementTypeTO(measurementType: MeasurementType, measurementTOs: List<MeasurementTO>): MeasurementTypeTO {
        val measurementTypeTO = MeasurementTypeTO()

        measurementTypeTO.id = measurementType.id
        measurementTypeTO.name = measurementType.name
        measurementTypeTO.displayName = measurementType.displayName
        measurementTypeTO.measurements = measurementTOs
        measurementTypeTO.thresholds = getThresholdTO(measurementType)

        return measurementTypeTO
    }

    fun getThresholdTO(measurementType: MeasurementType): ThresholdTO {
        val thresholdTO = ThresholdTO()

        thresholdTO.acceptable = measurementType.acceptable
        thresholdTO.informative = measurementType.informative
        thresholdTO.alarm = measurementType.alarm

        return thresholdTO
    }

    fun getCountyTO(county: CountyRegion, measurementTypeTOs: List<MeasurementTypeTO>, dateTime: ZonedDateTime): CountyTO {
        val countyTO = CountyTO()
        val now = dateTime.withZoneSameInstant(ZoneId.of(ZONE_ID)).format(DATE_TIME_FORMATTER)

        countyTO.name = county.name
        countyTO.now = now
        countyTO.weatherType = if (null == county.weatherType) 0 else county.weatherType
        countyTO.weatherDegree = if (null == county.weatherDegree) 0 else Math.round(county.weatherDegree!!)

        countyTO.measurementTypes = measurementTypeTOs

        return countyTO
    }

    fun getCountyLightTO(county: CountyRegion): CountyLightTO {
        val countyLightTO = CountyLightTO()

        countyLightTO.name = county.name
        countyLightTO.id = county.id
        countyLightTO.voivodeship = county.superName

        return countyLightTO
    }
}// Utils class
